File: php/src/LombokAlgoritma/DataStructure/DisjointSet.php
<?php

// LombokAlgoritma — disjoint set / union-find (SPEC §8.3)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\DataStructure;

use LombokAlgoritma\AlgoException;

/** Union by rank + full path compression; `find` returns the (normative) root. */
final class DisjointSet
{
    /** @var array<int, int> */
    private array $parent;
    /** @var array<int, int> */
    private array $rank;
    private int $count;

    public function __construct(int $n)
    {
        $this->parent = $n > 0 ? range(0, $n - 1) : [];
        $this->rank = $n > 0 ? array_fill(0, $n, 0) : [];
        $this->count = max(0, $n);
    }

    /** @throws AlgoException OUT_OF_BOUNDS for an element outside [0, n) */
    public function find(int $x): int
    {
        if ($x < 0 || $x >= count($this->parent)) {
            throw new AlgoException(
                AlgoException::OUT_OF_BOUNDS,
                "Index {$x} out of bounds [0, " . count($this->parent) . ')',
            );
        }
        $root = $x;
        while ($this->parent[$root] !== $root) {
            $root = $this->parent[$root];
        }
        while ($this->parent[$x] !== $root) {
            $next = $this->parent[$x];
            $this->parent[$x] = $root;
            $x = $next;
        }
        return $root;
    }

    /** Merge the sets of x and y; false when already joined. */
    public function union(int $x, int $y): bool
    {
        $rx = $this->find($x);
        $ry = $this->find($y);
        if ($rx === $ry) {
            return false;
        }
        if ($this->rank[$rx] < $this->rank[$ry]) {
            $this->parent[$rx] = $ry;
        } elseif ($this->rank[$rx] > $this->rank[$ry]) {
            $this->parent[$ry] = $rx;
        } else {
            $this->parent[$ry] = $rx;
            $this->rank[$rx]++;
        }
        $this->count--;
        return true;
    }

    public function connected(int $x, int $y): bool
    {
        return $this->find($x) === $this->find($y);
    }

    /** Number of disjoint sets. */
    public function count(): int
    {
        return $this->count;
    }
}

File: rust/lombokalgoritma/src/datastructure/disjoint_set.rs
// LombokAlgoritma — disjoint set / union–find (SPEC §8.3)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use crate::{Error, Result};
use alloc::vec;
use alloc::vec::Vec;

/// Union–find with union by rank and full path compression (near O(1) amortised).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DisjointSet {
    parent: Vec<usize>,
    rank: Vec<u8>,
    count: usize,
}

impl DisjointSet {
    /// `n` singleton sets `{0} … {n − 1}`.
    pub fn new(n: usize) -> Self {
        DisjointSet {
            parent: (0..n).collect(),
            rank: vec![0; n],
            count: n,
        }
    }

    /// Number of elements.
    pub fn len(&self) -> usize {
        self.parent.len()
    }

    /// `true` when there are no elements.
    pub fn is_empty(&self) -> bool {
        self.parent.is_empty()
    }

    /// Representative (root) of `x`'s set, compressing the whole path.
    ///
    /// # Errors
    /// [`Error::OutOfBounds`] when `x ≥ n`.
    pub fn find(&mut self, x: usize) -> Result<usize> {
        if x >= self.parent.len() {
            return Err(Error::OutOfBounds);
        }
        let mut root = x;
        while self.parent[root] != root {
            root = self.parent[root];
        }
        let mut cur = x;
        while self.parent[cur] != root {
            let next = self.parent[cur];
            self.parent[cur] = root;
            cur = next;
        }
        Ok(root)
    }

    /// Merge the sets of `x` and `y`; `false` when already in the same set.
    ///
    /// # Errors
    /// [`Error::OutOfBounds`] when `x` or `y` is `≥ n`.
    pub fn union(&mut self, x: usize, y: usize) -> Result<bool> {
        let rx = self.find(x)?;
        let ry = self.find(y)?;
        if rx == ry {
            return Ok(false);
        }
        match self.rank[rx].cmp(&self.rank[ry]) {
            core::cmp::Ordering::Less => self.parent[rx] = ry,
            core::cmp::Ordering::Greater => self.parent[ry] = rx,
            core::cmp::Ordering::Equal => {
                self.parent[ry] = rx;
                self.rank[rx] += 1;
            }
        }
        self.count -= 1;
        Ok(true)
    }

    /// `true` when `x` and `y` are in the same set.
    ///
    /// # Errors
    /// [`Error::OutOfBounds`] when `x` or `y` is `≥ n`.
    pub fn connected(&mut self, x: usize, y: usize) -> Result<bool> {
        Ok(self.find(x)? == self.find(y)?)
    }

    /// Number of disjoint sets.
    pub fn count(&self) -> usize {
        self.count
    }
}
